server/channel.go

```diff
@@ -0,0 +1,206 @@
+// pmm-agent
+// Copyright (C) 2018 Percona LLC
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+// Package server contains business logic of working with pmm-managed.
+package server
+
+import (
+	"fmt"
+	"sync"
+	"sync/atomic"
+
+	"github.com/percona/pmm/api/agent"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	serverRequestsCap = 32
+)
+
+// Channel encapsulates two-way communication channel between pmm-managed and pmm-agent.
+//
+// All exported methods are thread-safe.
+type Channel struct {
+	s agent.Agent_ConnectClient
+	l *logrus.Entry
+
+	lastSentRequestID uint32
+
+	sendM sync.Mutex
+
+	m         sync.Mutex
+	responses map[uint32]chan agent.ServerMessagePayload
+	requests  chan *agent.ServerMessage
+
+	closeOnce sync.Once
+	closeWait chan struct{}
+	closeErr  error
+}
+
+// NewChannel creates new two-way communication channel with given stream.
+//
+// Stream should not be used by the caller after channel is created.
+func NewChannel(stream agent.Agent_ConnectClient) *Channel {
+	s := &Channel{
+		s:         stream,
+		l:         logrus.WithField("component", "channel"), // only for debug logging
+		responses: make(map[uint32]chan agent.ServerMessagePayload),
+		requests:  make(chan *agent.ServerMessage, serverRequestsCap),
+		closeWait: make(chan struct{}),
+	}
+
+	go s.runReceiver()
+	return s
+}
+
+// close marks channel as closed with given error - only once.
+func (c *Channel) close(err error) {
+	c.closeOnce.Do(func() {
+		c.l.Debugf("Closing with error: %+v", err)
```

Comment:
shouldn't we close `requests` channel here?

Reply:
Only sender should close a channel.